packages/lesswrong/lib/sql/BulkWriter.ts

```diff
@@ -60,8 +60,8 @@ class BulkWriter<T extends DbObject> {
       this.queries.push(new InsertQuery(table, inserts.map(({document}) => document)));
     }
     if (updateOnes.length) {
-      this.queries = this.queries.concat(updateOnes.map(({filter, update, upsert}) => upsert
-        ? new InsertQuery(table, update as T, {}, {conflictStrategy: "upsert", upsertSelector: filter})
+      this.queries = this.queries.concat(updateOnes.map(({filter, update, upsert}) => upsert && update['$set']
+        ? new InsertQuery(table, update['$set'], {}, {conflictStrategy: "upsert", upsertSelector: filter})
         : new UpdateQuery(table, filter, update)
```

Comment:
I guess upserting wasn't used before. I think this is fine, I'm not sure if we should support anything other than `$set`, but let me know if you think otherwise. I also added tests for this in a file below.